In [18]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score
from sklearn.feature_extraction.text import TfidfVectorizer

In [2]:
df=pd.read_csv("climate_change_faqs.csv")
df.head()

source  \
0  https://www.ipcc.ch/site/assets/uploads/2020/0...   
1  https://www.ipcc.ch/site/assets/uploads/2020/0...   
2  https://www.ipcc.ch/site/assets/uploads/2020/0...   
3  https://www.ipcc.ch/site/assets/uploads/2020/0...   
4  https://www.ipcc.ch/site/assets/uploads/2020/0...   

                                                 faq text_type  
0  If Understanding of the Climate System Has Inc...         q  
1  The models used to calculate the IPCC’s temper...         a  
2               How Do We Know the World Has Warmed?         q  
3  Evidence for a warming world comes from multip...         a  
4   Have There Been Any Changes in Climate Extremes?         q

In [4]:
#Removing irrelavant features
df=df[["faq","text_type"]]
df=df.dropna(axis=0)
df.head()

faq text_type
0  If Understanding of the Climate System Has Inc...         q
1  The models used to calculate the IPCC’s temper...         a
2               How Do We Know the World Has Warmed?         q
3  Evidence for a warming world comes from multip...         a
4   Have There Been Any Changes in Climate Extremes?         q

In [21]:
# converting text_type into numeric value
df.text_type = pd.Categorical(df.text_type)
df['text_type'] = df.text_type.cat.codes
df.head()

faq  text_type
0  If Understanding of the Climate System Has Inc...          1
1  The models used to calculate the IPCC’s temper...          0
2               How Do We Know the World Has Warmed?          1
3  Evidence for a warming world comes from multip...          0
4   Have There Been Any Changes in Climate Extremes?          1

In [36]:
X=df["faq"]
Y=df["text_type"]
X=np.array(X)
Y=np.array(Y)

In [37]:
vectorizer=TfidfVectorizer()
X=vectorizer.fit_transform(X)
print("Dimension of original dataset",df.shape)
print("Dimension of dataset after tokenization",X.shape)

Dimension of original dataset (376, 2)
Dimension of dataset after tokenization (376, 6042)


In [41]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.1, random_state=42)

In [42]:
clf=RandomForestClassifier(max_depth=2,random_state=0)
clf.fit(X_train,Y_train)

RandomForestClassifier(max_depth=2, random_state=0)

In [43]:
#Y_test=Y_test.reshape(-1,1)
Y_pred=clf.predict(X_test)
f1_score(Y_test, Y_pred, average='macro')

0.9735191637630662

In [44]:
print("True Label: ",Y_test[0])
print("Predicted Label: ",Y_pred[0])

True Label:  1
Predicted Label:  1


The model gives F1-score of 0.97 which quite good